internal/shared/gocpy/gocpy.go

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+gocpy copies Go source code.
+The copied code is marked as generated.
+
+Usage:
+
+	gocpy [flags]
+
+The flags are:
+
+	-dest
+	    The destination filepath.
+	-pkg
+	    The destination package name (can contain comments).
+	-src
+	    The source filepath.
+*/
+package main
+
+import (
+	"bufio"
+	"flag"
+	"log"
+	"os"
+	"strings"
+)
+
+var (
+	dest = flag.String("dest", "", "The destination filepath.")
+	pkg  = flag.String("pkg", "", "The destination package name (can contain comments).")
+	src  = flag.String("src", "", "The source filepath.")
+)
+
+func main() {
+	flag.Parse()
+	if err := Copy(*dest, *pkg, *src); err != nil {
+		log.Fatalln(err)
+	}
+}
+
+// Copy copies a Go file from src to dst
+// with package renamed to pkg.
+func Copy(dest, pkg, src string) error {
+	inFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer inFile.Close()
+
+	outFile, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer outFile.Close()
+
+	w := bufio.NewWriter(outFile)
+	if _, err := w.WriteString("// Code generated by gocpy. DO NOT EDIT.\n\n"); err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(inFile)
+	for scanner.Scan() {
+		txt := scanner.Text()
+
+		if strings.HasPrefix(txt, "package ") {
+			if _, err := w.WriteString("package " + pkg + "\n"); err != nil {
```

Comment:
What about using `text/template` and making the input files templates?  That would allow replacing the package name and canonical import comment and eliminate the prefix check on each line.

Reply:
I find it easier to use actual code than writing Go source code templates. We get IDE, linting support when writing the shared code what we could not say about templates. We only need to replacing the package name which is not anything complex (I would argue that simpler than using templates.

Reply:
> If the performance is the concern, this could be easily improved by making sure that after the package name is replaced then we are not longer checking the line prefix. So the prefix check would be done only for the first couple of lines.

I did it in https://github.com/open-telemetry/opentelemetry-go-build-tools/pull/275